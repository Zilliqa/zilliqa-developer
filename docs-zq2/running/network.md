---
id: network
title: Network
keywords:
  - running
  - validator
description: Zilliqa 2.0 Network Instructions
---

TBD

