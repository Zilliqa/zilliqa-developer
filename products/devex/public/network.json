{
  "networks": [
    { "https://api.zilliqa.com": "Mainnet" },
    { "https://dev-api.zilliqa.com": "Testnet" },
    { "https://api.devnet.zilliqa.com": "Devnet" },
    { "https://zilliqa-isolated-server.zilliqa.com": "Isolated Server" }
  ]
}
