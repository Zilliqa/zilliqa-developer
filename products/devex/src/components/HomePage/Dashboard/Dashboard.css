.dashboard-container .card {
  border-radius: 10px;
}

.dashboard-container .card-header {
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
}

.dashboard-container .card-body {
  border-radius: 10px;
  box-shadow: 1px 2px 6px 7px rgba(154, 160, 185, 0.05), 0 15px 40px rgba(166, 173, 201, 0.0);
  padding: 1.25rem 1rem;
}