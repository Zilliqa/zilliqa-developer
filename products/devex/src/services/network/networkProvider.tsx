import React, { useState, useEffect, useContext } from "react";
import { useLocation, Link } from "react-router-dom";
import { Spinner } from "react-bootstrap";

import { DataService } from "./dataService";
import { UserPrefContext, NetworkMap } from "../userPref/userPrefProvider";

type NetworkState = {
  isIsolatedServer: boolean | null;
  dataService: DataService | null;
  networkUrl: string;
  apolloUrl: string;
  isValidUrl: boolean | null;
  inTransition: boolean;
  isLoadingNetworks: boolean;
};

export const QueryPreservingLink = ({
  to,
  style,
  className,
  onClick,
  children,
}: {
  to: string;
  style?: React.CSSProperties;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void;
  children: React.ReactNode;
}): JSX.Element => {
  const location = useLocation();
  return (
    <Link
      style={style}
      className={className}
      onClick={onClick}
      to={{
        pathname: to,
        search: location.search,
      }}
    >
      {children}
    </Link>
  );
};

export const useSearchParams = (): string => useLocation().pathname;

export const useNetworkUrl = (): string => {
  return new URLSearchParams(useLocation().search).get("network") || "";
};

export const useNetworkName = (): string => {
  const networkUrl = useNetworkUrl();
  const userPrefContext = useContext(UserPrefContext);

  if (!userPrefContext) {
    return "Loading ...";
  }
  const { networkMap } = userPrefContext;

  return (
    networkMap.get(networkUrl) || defaultNetworks.get(networkUrl) || networkUrl
  );
};

export let defaultNetworks: NetworkMap =
  process.env["REACT_APP_DEPLOY_ENV"] === "prod"
    ? new Map([
        ["https://api.zilliqa.com", "Mainnet"],
        ["https://dev-api.zilliqa.com", "Testnet"],
        ["https://api.devnet.zilliqa.com", "Devnet"],
        ["https://zilliqa-isolated-server.zilliqa.com", "Isolated Server"],
      ])
    : new Map([
        ["https://api.zilliqa.com", "Mainnet"],
        ["https://dev-api.zilliqa.com", "Testnet"],
        ["https://api.devnet.zilliqa.com", "Devnet"],
        ["https://zilliqa-isolated-server.zilliqa.com", "Isolated Server"],
        [
          "https://stg-zilliqa-isolated-server.zilliqa.com",
          "Staging Isolated Server",
        ],
      ]);

export const defaultApolloServers = new Map([
  ["Mainnet", "https://devex-apollo.zilliqa.com/"],
  ["Testnet", "https://stg-devex-apollo.zilliqa.com/"],
]);

export const useApolloServerAddress = (): string => {
  const networkName = useNetworkName();

  return (
    defaultApolloServers.get(networkName) ||
    defaultApolloServers.get("Mainnet") ||
    "http://localhost:5000"
  );
};

export const NetworkContext = React.createContext<NetworkState | null>(null);

export const NetworkProvider = (props: { children: any }) => {
  const userPrefContext = useContext(UserPrefContext);
  if (!userPrefContext) {
    return (
      <div className="center-spinner">
        <Spinner animation="border" />
      </div>
    );
  }
  const { networkMap, setNetworkMap } = userPrefContext;

  const network = useNetworkUrl();
  const apolloServer = useApolloServerAddress();

  const [state, setState] = useState<NetworkState>({
    isValidUrl: null,
    isIsolatedServer: null,
    dataService: null,
    networkUrl: network,
    apolloUrl: apolloServer,
    inTransition: true,
    isLoadingNetworks: true,
  });

  // Load optional networks from public folder
  useEffect(() => {
    let localNetworks;
    const loadNetworks = async () => {
      console.log("loading networks");
      try {
        const response = await fetch(process.env.PUBLIC_URL + "/networks.json");
        localNetworks = await response.json();
        defaultNetworks = new Map(
          localNetworks.networks.map(
            (x: { [url: string]: string }) => Object.entries(x)[0]
          )
        );
        if (networkMap.size === 0) setNetworkMap(defaultNetworks);
      } catch (e) {
        console.log("no local networks found");
        setNetworkMap(defaultNetworks);
      } finally {
        setState((prevState: NetworkState) => ({
          ...prevState,
          isLoadingNetworks: false,
        }));
      }
    };
    loadNetworks();
  }, []);

  useEffect(() => {
    if (state.isLoadingNetworks) return;
    return setState((prevState: NetworkState) => ({
      ...prevState,
      dataService: new DataService(network),
      inTransition: true,
      isIsolatedServer: null,
      networkUrl: network,
    }));
  }, [network, state.isLoadingNetworks]);

  // If dataservice changes, update isIsolatedServer
  useEffect(() => {
    let response: boolean;
    const checkNetwork = async () => {
      try {
        if (!state.dataService) return;
        response = await state.dataService.isIsolatedServer();
        setState((prevState: NetworkState) => ({
          ...prevState,
          isIsolatedServer: response,
          isValidUrl: true,
        }));
      } catch (e) {
        console.log(e);
        setState((prevState: NetworkState) => ({
          ...prevState,
          isValidUrl: false,
        }));
      } finally {
        setState((prevState: NetworkState) => ({
          ...prevState,
          inTransition: false,
        }));
      }
    };
    checkNetwork();
  }, [state.dataService]);

  return (
    <NetworkContext.Provider value={state}>
      {props.children}
    </NetworkContext.Provider>
  );
};
