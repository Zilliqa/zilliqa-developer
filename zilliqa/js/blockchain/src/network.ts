//  Copyright (C) 2018 Zilliqa
//
//  This file is part of zilliqa-js
//
//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <https://www.gnu.org/licenses/>.

import {
  Provider,
  RPCMethod,
  RPCResponse,
  ZilliqaModule,
} from "@zilliqa-js/core";
import { Wallet } from "@zilliqa-js/account";

export class Network implements ZilliqaModule {
  provider: Provider;
  signer: Wallet;

  constructor(provider: Provider, signer: Wallet) {
    this.provider = provider;
    this.signer = signer;
  }

  getNetworkId(): Promise<RPCResponse<string, string>> {
    return this.provider.send(RPCMethod.GetNetworkId);
  }

  getVersion(): Promise<RPCResponse<string, string>> {
    return this.provider.send(RPCMethod.GetVersion);
  }

  getNodeType(): Promise<RPCResponse<string, string>> {
    return this.provider.send(RPCMethod.GetNodeType);
  }

  getNumPeers(): Promise<RPCResponse<number, string>> {
    return this.provider.send(RPCMethod.GetNumPeers);
  }
}
