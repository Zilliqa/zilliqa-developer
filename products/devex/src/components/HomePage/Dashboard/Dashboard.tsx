import React, { useContext } from "react";
import { Container, Row, Col } from "react-bootstrap";

import { NetworkContext } from "src/services/network/networkProvider";
import { Spinner } from "react-bootstrap";

import BCInfo from "./BCInfo/BCInfo";
import DSBlockList from "./DSBlockList/DSBlockList";
import TxBlockList from "./TxBlockList/TxBlockList";
import ValTxnList from "./ValTxnList/ValTxnList";
// import PendTxnList from './PendTxnList/PendTxnList'

import "./Dashboard.css";
import ISInfo from "./ISInfo/ISInfo";

/*
              Dashboard Layout
    +++++++++++++++++++++++++++++++++++++
    |           BC Information          |
    +++++++++++++++++++++++++++++++++++++
    |  DS Blocks List | Tx Blocks List  |
    +++++++++++++++++++++++++++++++++++++
    |            ValTxnList             |
    +++++++++++++++++++++++++++++++++++++
    |            PendTxnList            |
    +++++++++++++++++++++++++++++++++++++
*/
const Dashboard: React.FC = () => {
  const networkContext = useContext(NetworkContext);
  if (!networkContext) {
    return (
      <div className="center-spinner">
        <Spinner animation="border" />
      </div>
    );
  }

  const { isIsolatedServer } = networkContext;

  return (
    <div>
      {isIsolatedServer ? (
        <>
          <Container className="dashboard-container">
            <Row>
              <ISInfo />
            </Row>
            <Row>
              <Col className="p-0">
                <ValTxnList />
              </Col>
            </Row>
          </Container>
        </>
      ) : (
        <>
          <Container className="dashboard-container">
            <Row>
              <BCInfo />
            </Row>
            <Row>
              <Col className="p-0" sm={5} md={5} lg={5}>
                <DSBlockList />
              </Col>
              <Col className="p-0 ml-4">
                <TxBlockList />
              </Col>
            </Row>
            <Row className="mt-3">
              <Col className="p-0">
                <ValTxnList />
              </Col>
            </Row>
            {/* <Row className='mt-3'>
              <Col className='p-0'>
                <PendTxnList />
              </Col>
            </Row> */}
          </Container>
        </>
      )}
    </div>
  );
};

export default Dashboard;
