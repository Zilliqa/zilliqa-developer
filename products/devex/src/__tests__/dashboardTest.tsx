import * as React from 'react'
import { mount } from 'enzyme'
import Dashboard from 'src/components/HomePage/Dashboard/Dashboard'
import { BrowserRouter as Router } from 'react-router-dom'
import { NetworkContext } from 'src/services/network/networkProvider'

describe('Isolated Server <Dashboard />', () => {

  const dashboard = mount(
    <Router>
      <NetworkContext.Provider value={{
        isValidUrl: null,
        isIsolatedServer: true,
        dataService: null,
        nodeUrl: '',
        inTransition: true,
        isLoadingUrls: true
      }}>
        <Dashboard />
      </NetworkContext.Provider>
    </Router>
  )

  it('matches snapshot', () => {
    expect(dashboard).toMatchSnapshot()
  })

})

describe('Non-Isolated Server <Dashboard />', () => {

  const dashboard = mount(
    <Router>
      <NetworkContext.Provider value={{
        isValidUrl: null,
        isIsolatedServer: false,
        dataService: null,
        nodeUrl: '',
        inTransition: true,
        isLoadingUrls: true
      }}>
        <Dashboard />
      </NetworkContext.Provider>
    </Router>
  )

  it('matches snapshot', () => {
    expect(dashboard).toMatchSnapshot()
  })

})