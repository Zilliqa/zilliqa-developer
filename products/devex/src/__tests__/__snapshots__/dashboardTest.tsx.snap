// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Isolated Server <Dashboard /> matches snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Dashboard>
      <div>
        <Container
          className="dashboard-container"
          fluid={false}
        >
          <div
            className="dashboard-container container"
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <ISInfo>
                  <Container
                    className="p-0"
                    fluid={false}
                  >
                    <div
                      className="p-0 container"
                    >
                      <Row
                        noGutters={false}
                      >
                        <div
                          className="row"
                        >
                          <Col>
                            <div
                              className="col"
                            >
                              <Card
                                body={false}
                                className="isinfo-card"
                              >
                                <div
                                  className="isinfo-card card"
                                >
                                  <CardBody>
                                    <div
                                      className="card-body"
                                    >
                                      <div>
                                        <Spinner
                                          animation="border"
                                          role="status"
                                        >
                                          <div
                                            className="spinner-border"
                                            role="status"
                                          />
                                        </Spinner>
                                      </div>
                                    </div>
                                  </CardBody>
                                </div>
                              </Card>
                            </div>
                          </Col>
                          <Col>
                            <div
                              className="col"
                            >
                              <Card
                                body={false}
                                className="isinfo-card"
                              >
                                <div
                                  className="isinfo-card card"
                                >
                                  <CardBody>
                                    <div
                                      className="card-body"
                                    >
                                      <div>
                                        <Spinner
                                          animation="border"
                                          role="status"
                                        >
                                          <div
                                            className="spinner-border"
                                            role="status"
                                          />
                                        </Spinner>
                                      </div>
                                    </div>
                                  </CardBody>
                                </div>
                              </Card>
                            </div>
                          </Col>
                        </div>
                      </Row>
                    </div>
                  </Container>
                </ISInfo>
              </div>
            </Row>
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Col
                  className="p-0"
                >
                  <div
                    className="p-0 col"
                  >
                    <ValTxnList>
                      <Card
                        body={false}
                        className="valtxlist-card"
                      >
                        <div
                          className="valtxlist-card card"
                        >
                          <CardHeader>
                            <div
                              className="card-header"
                            >
                              <div
                                className="valtxlist-card-header"
                              >
                                <span>
                                  Transactions
                                </span>
                                <QueryPreservingLink
                                  to="/tx"
                                >
                                  <Link
                                    to={
                                      Object {
                                        "pathname": "/tx",
                                        "search": "",
                                      }
                                    }
                                  >
                                    <LinkAnchor
                                      href="/tx"
                                      navigate={[Function]}
                                    >
                                      <a
                                        href="/tx"
                                        onClick={[Function]}
                                      >
                                        View Recent Transactions
                                      </a>
                                    </LinkAnchor>
                                  </Link>
                                </QueryPreservingLink>
                              </div>
                            </div>
                          </CardHeader>
                          <CardBody>
                            <div
                              className="card-body"
                            >
                              <Spinner
                                animation="border"
                                role="status"
                              >
                                <div
                                  className="spinner-border"
                                  role="status"
                                />
                              </Spinner>
                            </div>
                          </CardBody>
                        </div>
                      </Card>
                    </ValTxnList>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </Container>
      </div>
    </Dashboard>
  </Router>
</BrowserRouter>
`;

exports[`Non-Isolated Server <Dashboard /> matches snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Dashboard>
      <div>
        <Container
          className="dashboard-container"
          fluid={false}
        >
          <div
            className="dashboard-container container"
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <BCInfo>
                  <Card
                    body={false}
                    className="bcstats-card"
                  >
                    <div
                      className="bcstats-card card"
                    >
                      <CardBody>
                        <div
                          className="card-body"
                        >
                          <Spinner
                            animation="border"
                            role="status"
                          >
                            <div
                              className="spinner-border"
                              role="status"
                            />
                          </Spinner>
                        </div>
                      </CardBody>
                    </div>
                  </Card>
                </BCInfo>
              </div>
            </Row>
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Col
                  className="p-0"
                  lg={5}
                  md={5}
                  sm={5}
                >
                  <div
                    className="p-0 col-lg-5 col-md-5 col-sm-5"
                  >
                    <DSBlockList>
                      <Card
                        body={false}
                        className="dsblock-card"
                      >
                        <div
                          className="dsblock-card card"
                        >
                          <CardHeader>
                            <div
                              className="card-header"
                            >
                              <div
                                className="dsblock-card-header"
                              >
                                <span>
                                  DS Blocks
                                </span>
                                <QueryPreservingLink
                                  to="/dsbk"
                                >
                                  <Link
                                    to={
                                      Object {
                                        "pathname": "/dsbk",
                                        "search": "",
                                      }
                                    }
                                  >
                                    <LinkAnchor
                                      href="/dsbk"
                                      navigate={[Function]}
                                    >
                                      <a
                                        href="/dsbk"
                                        onClick={[Function]}
                                      >
                                        View All
                                      </a>
                                    </LinkAnchor>
                                  </Link>
                                </QueryPreservingLink>
                              </div>
                            </div>
                          </CardHeader>
                          <CardBody>
                            <div
                              className="card-body"
                            >
                              <Spinner
                                animation="border"
                                role="status"
                              >
                                <div
                                  className="spinner-border"
                                  role="status"
                                />
                              </Spinner>
                            </div>
                          </CardBody>
                        </div>
                      </Card>
                    </DSBlockList>
                  </div>
                </Col>
                <Col
                  className="p-0 ml-4"
                >
                  <div
                    className="p-0 ml-4 col"
                  >
                    <TxBlockList>
                      <Card
                        body={false}
                        className="txblock-card"
                      >
                        <div
                          className="txblock-card card"
                        >
                          <CardHeader>
                            <div
                              className="card-header"
                            >
                              <div
                                className="dsblock-card-header"
                              >
                                <span>
                                  Transaction Blocks
                                </span>
                                <QueryPreservingLink
                                  to="/txbk"
                                >
                                  <Link
                                    to={
                                      Object {
                                        "pathname": "/txbk",
                                        "search": "",
                                      }
                                    }
                                  >
                                    <LinkAnchor
                                      href="/txbk"
                                      navigate={[Function]}
                                    >
                                      <a
                                        href="/txbk"
                                        onClick={[Function]}
                                      >
                                        View All
                                      </a>
                                    </LinkAnchor>
                                  </Link>
                                </QueryPreservingLink>
                              </div>
                            </div>
                          </CardHeader>
                          <CardBody>
                            <div
                              className="card-body"
                            >
                              <Spinner
                                animation="border"
                                role="status"
                              >
                                <div
                                  className="spinner-border"
                                  role="status"
                                />
                              </Spinner>
                            </div>
                          </CardBody>
                        </div>
                      </Card>
                    </TxBlockList>
                  </div>
                </Col>
              </div>
            </Row>
            <Row
              className="mt-3"
              noGutters={false}
            >
              <div
                className="mt-3 row"
              >
                <Col
                  className="p-0"
                >
                  <div
                    className="p-0 col"
                  >
                    <ValTxnList>
                      <Card
                        body={false}
                        className="valtxlist-card"
                      >
                        <div
                          className="valtxlist-card card"
                        >
                          <CardHeader>
                            <div
                              className="card-header"
                            >
                              <div
                                className="valtxlist-card-header"
                              >
                                <span>
                                  Transactions
                                </span>
                                <QueryPreservingLink
                                  to="/tx"
                                >
                                  <Link
                                    to={
                                      Object {
                                        "pathname": "/tx",
                                        "search": "",
                                      }
                                    }
                                  >
                                    <LinkAnchor
                                      href="/tx"
                                      navigate={[Function]}
                                    >
                                      <a
                                        href="/tx"
                                        onClick={[Function]}
                                      >
                                        View Recent Transactions
                                      </a>
                                    </LinkAnchor>
                                  </Link>
                                </QueryPreservingLink>
                              </div>
                            </div>
                          </CardHeader>
                          <CardBody>
                            <div
                              className="card-body"
                            >
                              <Spinner
                                animation="border"
                                role="status"
                              >
                                <div
                                  className="spinner-border"
                                  role="status"
                                />
                              </Spinner>
                            </div>
                          </CardBody>
                        </div>
                      </Card>
                    </ValTxnList>
                  </div>
                </Col>
              </div>
            </Row>
            <Row
              className="mt-3"
              noGutters={false}
            >
              <div
                className="mt-3 row"
              >
                <Col
                  className="p-0"
                >
                  <div
                    className="p-0 col"
                  >
                    <PendTxnList>
                      <Card
                        body={false}
                        className="pendtxlist-card"
                      >
                        <div
                          className="pendtxlist-card card"
                        >
                          <CardHeader>
                            <div
                              className="card-header"
                            >
                              <div
                                className="pendtxlist-card-header"
                              >
                                <span>
                                  Pending Transactions
                                </span>
                              </div>
                            </div>
                          </CardHeader>
                          <CardBody>
                            <div
                              className="card-body"
                            >
                              <Spinner
                                animation="border"
                                role="status"
                              >
                                <div
                                  className="spinner-border"
                                  role="status"
                                />
                              </Spinner>
                            </div>
                          </CardBody>
                        </div>
                      </Card>
                    </PendTxnList>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </Container>
      </div>
    </Dashboard>
  </Router>
</BrowserRouter>
`;
